src/store/Controller.cc

```diff
@@ -296,6 +296,14 @@ Store::Controller::dereferenceIdle(StoreEntry &e, bool wantsLocalMemory)
             keepInStoreTable = wantsLocalMemory || keepInStoreTable;
     }
 
+    if (e.hittingRequiresCollapsing()) {
+        // If we were writing this now-locally-idle entry, then we did not
+        // finish and should now destroy an incomplete entry. Otherwise, do not
+        // leave this idle StoreEntry behind because handleIMSReply() lacks
+        // freshness checks when hitting a collapsed revalidation entry.
+        keepInStoreTable = false; // may overrule fs decisions made above
+    }
```

Comment:
I would move this block above, just below 'ENTRY_SPECIAL' check.

Reply:
We cannot move this block up -- the "fs" code needs to be notified about this entry _and_ may decide to keep it in store_table. I hinted at that order by referencing "fs decisions made above" in the assignment comment. There is also a `Notify the fs` comment earlier in the method.Perhaps those hints are too subtle. Does the comment adjustment in 8ec3f30 address your concern?

Reply:
> We cannot move this block up

I think you are right. I hoped that we could somehow refactor the code for the path when e.hittingRequiresCollapsing() is true - in this case, keepInStoreTable flag should not be recalculated. But it is difficult to do that since dereference() methods are responsible for two tasks at once - 'dereferencing' and returning a boolean constant (false for Rock and MemStore and true for Ufs). This refactoring would require splitting dereference() into two methods (e.g., dereference() and keepInStoreTable()). I do not think it is worth doing here.

Reply:
We are on the same page then, thank you.